ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/PeerSimulator/Handlers.hs

```diff
@@ -0,0 +1,136 @@
+{-# LANGUAGE LambdaCase #-}
+
+-- | Business logic of the SyncChain protocol handlers that operates
+-- on the 'AdvertisedPoints' of a point schedule.
+--
+-- These are separated from the scheduling related mechanics of the
+-- ChainSync server mock that the peer simulator uses, in
+-- "Test.Consensus.PeerSimulator.ScheduledChainSyncServer".
+module Test.Consensus.PeerSimulator.Handlers (
+    handlerFindIntersection
+  , handlerRequestNext
+  ) where
+
+import           Control.Monad.Trans (lift)
+import           Control.Monad.Writer.Strict (MonadWriter (tell),
+                     WriterT (runWriterT))
+import           Data.Coerce (coerce)
+import           Data.Maybe (fromJust)
+import           Data.Monoid (First (..))
+import           Ouroboros.Consensus.Block.Abstract (Point (..), getHeader)
+import           Ouroboros.Consensus.Util.Condense (Condense (..))
+import           Ouroboros.Consensus.Util.IOLike (IOLike, STM, StrictTVar,
+                     readTVar, writeTVar)
+import           Ouroboros.Network.AnchoredFragment (AnchoredFragment)
+import qualified Ouroboros.Network.AnchoredFragment as AF
+import           Ouroboros.Network.Block (blockPoint, getTipPoint)
+import qualified Test.Consensus.BlockTree as BT
+import           Test.Consensus.BlockTree (BlockTree)
+import           Test.Consensus.PeerSimulator.ScheduledChainSyncServer
+                     (FindIntersect (..),
+                     RequestNext (AwaitReply, RollBackward, RollForward))
+import           Test.Consensus.PointSchedule (AdvertisedPoints (header, tip),
+                     HeaderPoint (HeaderPoint), TipPoint (TipPoint))
+import           Test.Util.Orphans.IOLike ()
+import           Test.Util.TestBlock (TestBlock)
+
+-- | Find the first fragment contained in the first arg that starts at one of the given points.
+intersectWith :: AnchoredFragment TestBlock -> [Point TestBlock] -> Maybe (Point TestBlock)
```

Comment:
Is this function specific to `AnchoredFragment TestsBlock`? If not perhaps it's worth adding it to the the same module where `AnchoredFragment` is defined.